### PR TITLE
cache: implement the circuit breaker pattern for asynchronous set operations in the cache client

### PR DESCRIPTION
**Pull request by damnever**
Ref https://github.com/thanos-io/thanos/issues/6975

* [x] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

## Changes

Wrap SetAsync with a circuit breaker to dynamically limit the cache backfilling requests. Both the memcached and redis clients have added the same circuit breaker configuration.

## Verification

Added some unit tests.

### CODE REVIEW

File: pkg/cacheutil/memcached_client.go
Comment (yeya24) [inline pkg/cacheutil/memcached_client.go:188]: Since we enable the circuit breaker by default, is there any downside to do so? Do you think it is worth it to put it behind a FF?

File: pkg/cacheutil/memcached_client_test.go
Comment (yeya24) [inline pkg/cacheutil/memcached_client_test.go:568]: It might be fine in a unit test... But do we need lock here? Default memcached client config has async concurrency > 1

File: pkg/cacheutil/memcached_client.go
Comment (yeya24) [inline pkg/cacheutil/memcached_client.go:64]: We might need to keep an eye on this to make sure those values sane.
So far, it looks good to me. Would be better to have another pair of eyes on it

**Review by yeya24 — COMMENTED** [review]
Overall it is great work and it makes a lot of sense. Thanks.
We need to update docs.

**Review by GiedriusS — COMMENTED** [review]
Please update docs with new fields + add a paragraph explaining how this works :smile:

File: pkg/cacheutil/memcached_client.go
Comment (damnever) [reply pkg/cacheutil/memcached_client.go:202]: feature flag has been added

File: pkg/cacheutil/memcached_client_test.go
Comment (damnever) [reply pkg/cacheutil/memcached_client_test.go:568]: I think the default `lock` is sufficient, as nothing else accesses this variable.

Comment (yeya24) [reply pkg/cacheutil/memcached_client_test.go:568]: I don't know why I didn't see that lock. All good then

**Review by yeya24 — DISMISSED** [review]
Thanks!

**Comment by yeya24** [comment]
```
17:08:51 redis-redis: 1:C 26 Jan 2024 17:08:51.802 # Configuration loaded
17:08:51 redis-redis: 1:M 26 Jan 2024 17:08:51.802 * monotonic clock: POSIX clock_gettime
17:08:51 redis-redis: 1:M 26 Jan 2024 17:08:51.803 * Running mode=standalone, port=6379.
17:08:51 redis-redis: 1:M 26 Jan 2024 17:08:51.803 # Server initialized
17:08:51 redis-redis: 1:M 26 Jan 2024 17:08:51.803 # WARNING overcommit_memory is set to 0! Background save may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
17:08:51 redis-redis: 1:M 26 Jan 2024 17:08:51.803 * Ready to accept connections
17:08:52 static: 172.30.0.8 - - [26/Jan/2024:17:08:52 +0000] "GET /metrics HTTP/1.1" 200 172 "-" "Prometheus/2.41.0" "-"
17:08:52 Ports for container redis-client-redis-redis >> Local ports: map[redis:6379] Ports available from host: map[redis:32919]
    store_gateway_test.go:1001: store_gateway_test.go:1001: ""
        
         unexpected error: set async circuit breaker: consecutive failures must be greater than 0
        
17:08:52 Killing redis-redis
=== CONT  TestRangeQueryShardingWithRandomData
--- FAIL: TestRedisClient_Rueidis (2.90s)
```

We need to fix E2E test. @damnever  https://github.com/thanos-io/thanos/actions/runs/7665969811/job/20908734057?pr=7010

**Comment by yeya24** [comment]
@damnever There is one more lint to fix before we can merge. Thanks

**Review by yeya24 — APPROVED** [review]

**Comment by yeya24** [comment]
@GiedriusS Wanna take another look before I merge this code change?

File: pkg/cacheutil/redis_client.go
Comment (GiedriusS) [inline pkg/cacheutil/redis_client.go:277]: Should this be factored out into another function? I think it's exactly the same in `memcached_client`.

File: docs/components/store.md
Comment (GiedriusS) [inline docs/components/store.md:351]: `If set to 0, the circuit breaker resets it to 60 seconds.`

What does it mean? That the default value of this is `60 seconds`? :thinking:

Comment (damnever) [reply docs/components/store.md:351]: Yes, github.com/sony/gobreaker uses this default value.

File: pkg/cacheutil/redis_client.go
Comment (damnever) [reply pkg/cacheutil/redis_client.go]: I have separated the entire configuration and logic, including this function. Please take another look.

**Comment by yeya24** [comment]
E2E test failure

```
=== NAME  TestRuleNativeHistograms
    native_histograms_test.go:391: native_histograms_test.go:391: ""
        
         unexpected error: unable to find metrics [prometheus_remote_storage_histograms_total] with expected values after 50 retries. Last error: <nil>. Last values: [0]
```

Doesn't seem like related. Ignoring for now

**Comment by yeya24** [comment]
Hey @GiedriusS, I will merge this pr now. If you find anything else that's worth improving, feel free to create a separate issue and we can address later.

**Comment by GiedriusS** [comment]
:+1: I am also thinking about whether it makes sense to only protect SET operations through the circuit breaker pattern? I mean if the cache is down then GETs won't work either.

**Comment by damnever** [comment]
I agree; I think GET is more sensitive to latency and should be assigned a higher priority.

**Review by GiedriusS — COMMENTED** [review]
So maybe let's remove the `set_` prefix from config and put gets under the circuit breaker too? Should be a small change. Ideally this would be done before the next release

**Comment by damnever** [comment]
Removing the `set_` prefix sounds fine, but I recommend using a different circuit breaker for the GET operation. The intent of this PR is to try to limit the resources used by async set operations(#6975), subsequently benefiting other operations such as GETs.

**Pull request merged**
